### ISSUE TITLE
[🐛 Bug]: Cannot set read timeout value in Local Drivers

### ISSUE THREAD

**Issue by TolstoyDotCom** [SeleniumHQ/selenium#11158]
### What happened?

Despite setting high timeouts ("{implicit: 10000000, pageLoad: 10000000, script: 10000000}"), calling executeAsyncScript will occasionally time out after exactly 3 minutes. This is using 4.5.0 of Selenium and Java 17 with Firefox 105 on Ubuntu 18. It's also using selenium-http-jdk-client.

### How can we reproduce the issue?

```shell
Download https://github.com/TolstoyDotCom/more-speech and follow the instructions in the "2.1 Notes" section. You might need to use different handles and different numbers of pages to see the issue.

When you click 'Check timeline', the app gets some tweets from the user's profile page then visits each tweet's page.

On an individual tweet page, com.tolstoy.censorship.twitter.checker.app.webdriver.WebDriverFactoryNT first calls executeAsyncScript to load the tweets using the scripts in the stockscripts directory (jquery + suededenim + tweet_retriever).

Then, com.tolstoy.censorship.twitter.checker.app.browserproxy.BrowserProxyHAR calls executeAsyncScript using (jquery + suededenim + har_retriever).

Then, WebDriverFactoryNT calls executeAsyncScript using (jquery + suededenim + json_parser).

So, three calls to executeAsyncScript per page and if one of them takes longer than 3 minutes (likely if there are lots of tweets), it will probably time out.
```

### Relevant log output

```shell
[INFO ] 2022-Oct-21 11:45:18 AM c.t.c.t.c.a.h.SearchRunTimelineBuilder - Loading reply page [...]
[...]
[ERROR] 2022-Oct-21 11:48:18 AM c.t.c.t.c.a.h.SearchRunTimelineBuilder - error logging in or building searchRun org.openqa.selenium.TimeoutException: java.util.concurrent.TimeoutException
Build info: version: '4.5.0', revision: 'fe167b119a'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-125-generic', java.version: '17.0.4'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Command: [c57d3478-3d94-41ae-a991-f4792364f495, get {url=https://twitter.com/[...]}]
Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 105.0, moz:accessibilityChecks: false, moz:buildID: 20220915150737, moz:debuggerAddress: 127.0.0.1:36466, moz:geckodriverVersion: 0.32.0, moz:headless: false, moz:platformVersion: 5.4.0-125-generic, moz:processID: 6398, moz:profile: /tmp/rust_mozprofileAhZ2z2, moz:shutdownTimeout: 60000, moz:useNonSpecCompliantPointerOrigin: false, moz:webdriverClick: true, moz:windowless: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(), se:cdp: ws://127.0.0.1:36466/devtoo..., se:cdpVersion: 85.0, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 10000000, pageLoad: 10000000, script: 10000000}, unhandledPromptBehavior: dismiss}
Session ID: c57d3478-3d94-41ae-a991-f4792364f495
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.makeCall(NettyHttpHandler.java:65)
	at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:56)
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.execute(NettyHttpHandler.java:49)
	at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:56)
	at org.openqa.selenium.remote.http.netty.NettyClient.execute(NettyClient.java:99)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:181)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:167)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:142)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:547)
	at org.openqa.selenium.remote.RemoteWebDriver.get(RemoteWebDriver.java:310)
	at com.tolstoy.censorship.twitter.checker.app.helpers.SearchRunTimelineBuilder.getIndividualPage(SearchRunTimelineBuilder.java:353)
	at com.tolstoy.censorship.twitter.checker.app.helpers.SearchRunTimelineBuilder.getIndividualPages(SearchRunTimelineBuilder.java:323)
	at com.tolstoy.censorship.twitter.checker.app.helpers.SearchRunTimelineBuilder.buildSearchRunTimelineInternal(SearchRunTimelineBuilder.java:282)
	at com.tolstoy.censorship.twitter.checker.app.helpers.SearchRunTimelineBuilder.buildSearchRunTimeline(SearchRunTimelineBuilder.java:185)
	at com.tolstoy.censorship.twitter.checker.app.AppGUI$TimelineWorker.doInBackground(AppGUI.java:210)
	at com.tolstoy.censorship.twitter.checker.app.AppGUI$TimelineWorker.doInBackground(AppGUI.java:187)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at org.asynchttpclient.netty.NettyResponseFuture.get(NettyResponseFuture.java:206)
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.makeCall(NettyHttpHandler.java:59)
	... 23 more
```

### Operating System

Ubuntu 18

### Selenium version

4.5.0 Java

### What are the browser(s) and version(s) where you see this issue?

Firefox 105

### What are the browser driver(s) and version(s) where you see this issue?

webdrivermanager 5.0.3

### Are you using Selenium Grid?

no

**Comment by diemol** [comment]
Can you provide a reduced test case where this can be reproduced?

**Comment by titusfortner** [comment]
The timeout you are getting has nothing to do with the WebDriver specified timeouts.

You need to update the Java http library read timeout (default is 3 minutes):

```
ClientConfig config = ClientConfig.defaultConfig().readTimeout(Duration.ofSeconds(300));
RemoteWebDriver driver = ChromeDriver.builder()
                .address(myHub)
                .config(config)
                .oneOf(options)
                .build();
```

**Comment by TolstoyDotCom** [comment]
This was happening when I tested with other versions of Selenium that don't use selenium-http-jdk-client, such as version 4.1.1 of Selenium.

I'm also loading the browser on the user's computer, I'm not using grid or some other remote setup. This is a desktop application that loads the user's browser and navigates to Twitter pages to compile a list of tweets. It's just as if the user opened their browser, navigated a few Twitter pages, and recorded which tweets appear where.

The application uses Javascript to scroll the Twitter pages, scrape info from the HTML, read the HAR logs, and merge data. That's where the stack trace happened: after the code called executeAsyncScript with the Javascript code that scrolls a Twitter page and scrapes the HTML. So, it seems like this is a script timeout, not, for instance, a timeout when asking Selenium to load a page and the connection takes a long time.

I tried the solution above. In WebDriverFactoryNT.makeWebDriver() I changed this:

`WebDriver driver = new FirefoxDriver( firefoxOptions );`

to this:

```
ClientConfig clientConfig = ClientConfig.defaultConfig().readTimeout( TIMEOUT ).connectionTimeout( TIMEOUT );
WebDriver driver = FirefoxDriver.builder().config( clientConfig ).oneOf( firefoxOptions ).build();
```

That results in RemoteWebDriverBuilder.getLocalDriver() throwing "ClientConfig instances do not work for Local Drivers".

This is how I'm currently setting timeouts (TIMEOUT is a Duration of over 2 hours):

```
firefoxOptions.setPageLoadTimeout( TIMEOUT );
firefoxOptions.setScriptTimeout( TIMEOUT );
firefoxOptions.setImplicitWaitTimeout​( TIMEOUT );
[...]
driver.manage().timeouts().scriptTimeout( TIMEOUT );

```

Is there some other way?

**Comment by titusfortner** [comment]
All http libraries have default read timeout values.

Unfortunately, we currently do not have a way to change them in local drivers in Java right now.

I think we need to add a `ClientConfig` parameter to driver constructors to allow this.

**Comment by TolstoyDotCom** [comment]
I pushed a new version that doesn't show the whole of my script in the stack trace.

I was able to get timeout errors twice using that version with these settings:
handle to check: alexborstein
skip login: [checked]
number of timeline replies to check: 1
how many pages on the timeline to scroll: 1
firefox profile name: twtr_censorship

I'd already created that Firefox profile, logged in to a testing Twitter account, and enabled the HAR extension.

**Comment by TolstoyDotCom** [comment]
I'm checking in to see if there's any news on this issue. I'm not familiar with Selenium's internals, but if you can point me in the right direction I might be able to help.

Alternatively, if I created a local override of one of the libraries you use, what would I change in the library to get a longer script timeout?

**Comment by krmahadevan** [comment]
@titusfortner - Quick question.

The below sample (using `7.7.2`) works fine. In firefox it triggers some Javascript related errors (my JS skills aren't that great) but it doesn't trigger any sort of timeouts.

Wanted to know if we are looking at recommending to the users that they should be using the below sort of variants to alter the httpclient level timeouts (when working with local webdrivers) or are we looking at altering the constructors to accept a `ClientConfig` object for each of the webdriver constructors (I am pretty sure that this is not the case, but still asking)

```java
import java.time.Duration;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.GeckoDriverService;
import org.openqa.selenium.remote.http.ClientConfig;

public class SampleTestCase {

  public static void main(String[] args) {
    runUsingFirefox();
//    runUsingChrome();
  }

  private static void runUsingFirefox() {
    WebDriver driver = null;
    Duration timeout = Duration.ofMinutes(4);
    try {
      ClientConfig clientConfig = ClientConfig.defaultConfig()
          .readTimeout(timeout).connectionTimeout(timeout);
      driver = FirefoxDriver.builder()
          .withDriverService(GeckoDriverService.createDefaultService())
          .config(clientConfig).build();

      driver.manage().timeouts().scriptTimeout(Duration.ofSeconds(200));
      ((JavascriptExecutor) driver).executeScript(getScript());
    } finally {
      if (driver != null) {
        driver.quit();
      }
    }
  }

  private static void runUsingChrome() {
    WebDriver driver = null;
    Duration timeout = Duration.ofMinutes(4);
    try {
      ClientConfig clientConfig = ClientConfig.defaultConfig()
          .readTimeout(timeout).connectionTimeout(timeout);
      driver = ChromeDriver.builder()
          .withDriverService(ChromeDriverService.createDefaultService())
          .config(clientConfig).build();

      driver.manage().timeouts().scriptTimeout(Duration.ofSeconds(200));
      ((JavascriptExecutor) driver).executeScript(getScript());
    } finally {
      if (driver != null) {
        driver.quit();
      }
    }
  }

  private static String getScript() {
    return "function timeout(ms) { "
        + "return new Promise(resolve => setTimeout(resolve, ms)); "
        + "};"
        + "const printFunction = async () => {"
        + "await timeout(190000); "
        + "return \"bla bla\""
        + "};"
        + "await printFunction();";
  }
}

```

**Comment by titusfortner** [comment]
Ok, so this properly sets the read timeout to 4 seconds on a local driver!
```
        Duration timeout = Duration.ofSeconds(4);
        ClientConfig clientConfig = ClientConfig.defaultConfig()
                .readTimeout(timeout).connectionTimeout(timeout);
        WebDriver driver = ChromeDriver.builder()
                .withDriverService(ChromeDriverService.createDefaultService())
                .config(clientConfig)
                .build();

        driver.get("https://titusfortner.com/examples/never_loads.html");
```

There are 2 things I don't like about this.
1. What comes back isn't a `ChromeDriver` instance and it really should be, which means it gets augmented which it really shouldn't be.
2. It requires passing in a default argument to satisfy the toggle condition

The code is the exact same as:
```
        WebDriver driver = RemoteWebDriver.builder()
                .oneOf(new ChromeOptions())
                .withDriverService(ChromeDriverService.createDefaultService())
                .config(clientConfig)
                .build();
```

Maybe it is ok that it isn't a ChromeDriver instance?

If we don't care about it being a ChromeDriver instance, we could implement each driver with this so user only needs to specify values they care about:
```
  public static RemoteWebDriverBuilder builder() {
    return RemoteWebDriver.builder()
        .oneOf(new ChromeOptions())
        .withDriverService(ChromeDriverService.createDefaultService();
  }
```

The reason RemoteWebDriver can set http client info is because it has this constructor:
```
  public RemoteWebDriver(CommandExecutor executor, Capabilities capabilities) {
```

We would need to either make this constructor public in subclasses or provide an additional constructor that would allow changing values. 

To be honest, your suggestion is probably the best compromise since anything we change is going to have drawbacks.

**Comment by diemol** [comment]
I think it is OK it is not a `ChromeDriver` instance, as I have seen the need to set a timeout locally to be rare. I agree with the second point, passing a default value is not ideal. Would be nice to change that and add docs about setting a timeout for the remote and local use case.

**Comment by krmahadevan** [comment]
@titusfortner - 

> It requires passing in a default argument to satisfy the toggle condition

Are you referring to `.withDriverService(GeckoDriverService.createDefaultService())` ?

@titusfortner @diemol 

I was thinking of the following additional choices:

### Option 1

There are two variants in this:

1. Within `org.openqa.selenium.firefox.FirefoxDriver#builder` what if we by default add a `GeckoDriverService.createDefaultService()` so that it becomes like this (Same goes for the builders for the other browser drivers) [ _This is the same as what @titusfortner proposed_ ]

```java
public static RemoteWebDriverBuilder builder() {
    return RemoteWebDriver.builder()
               .withDriverService(GeckoDriverService.createDefaultService())
               .oneOf(new FirefoxOptions(), new Capabilities[0]);
}
```
This way we can avoid having to pass the default service explicitly.

**Disadvantage:** 
If user is using the builder to work with a standalone or a grid, then we end up breaking backward compatibility.

2. We add a new builder variant say `localBuilder()` and within it we add the driver service setting

```java
public static RemoteWebDriverBuilder localBuilder() {
    return RemoteWebDriver.builder()
               .withDriverService(GeckoDriverService.createDefaultService())
               .oneOf(new FirefoxOptions(), new Capabilities[0]);
}
```

### Option 2
We resort to letting the user define the read and connection timeouts via 

1.  Some file available in the CLASSPATH (for e.g., `http_client_config.properties` ) (or)
2. JVM arguments (or)
3. Environment variables

and we honour the values specified for them.

**Disadvantage:** 
A user would NOT be able to set these timeouts on a per `WebDriver` basis which they can do if they use the builder approach and it would be enabled at all the webdriver levels.

### Option 3

Add a new overloaded constructor for each of the `Driver` implementations to accept a `ClientConfig` object so that a user can easily setup the timeout.

**Disadvantage:**  _Personal opinion_

Given that we already have a builder for the webdriver instantiations, I think this only keeps adding to the constructor variants. A builder is perhaps the most easiest way to build a webdriver object easily and I feel that we should perhaps start encouraging our users to start looking at leveraging it instead of the webdriver constructor variants.

**Comment by pujagani** [comment]
@krmahadevan Appreciate the detailed suggestions. It can probably be a discussion topic for tomorrow's TLC meeting to identify the best approach.

**Comment by krmahadevan** [comment]
@pujagani I added it to the google doc.

**Comment by diemol** [comment]
Adding my opinion here because I cannot join the meeting. 

I think Option 1 is still viable and backwards compatibility won't be broken if we change the internal logic in `RemoteWebDriverBuilder`. 

In short, the logic in `RemoteWebDriverBuilder` will favor a local driver if a driver service is set. However, we could flip that internal logic and favor a remote driver if a URI is set (either through the builder or through `ClientConfig`. With that, if a local driver is set and a remote host or URI are set, the logic should try to create a remote connection.

**Comment by pujagani** [comment]
+1 with @diemol's suggestion above.

**Comment by titusfortner** [comment]
@krmahadevan what is the advantage of the second one here? `.oneOf(new FirefoxOptions(), new Capabilities[0])`. If we're setting this on the subclassed Driver, should only need the first one as the default and it will get overridden if users want to do something else.

> If user is using the builder to work with a standalone or a grid, then we end up breaking backward compatibility.
It would only be used for `FirefoxDriver.builder()`, not `RemoteWebDriver.builder()` so I think it would be fine?

Option 2 — I *really dislike environment variable/system property toggles for things. Unless we have to do this for the grid, I really want us to avoid them.

Option 3 — My personal preference would favor this over option 2, but already the driver construction is complicated.

@diemol I'm not following what you are saying about preferring local or remote driver; this isn't true:
> the logic in RemoteWebDriverBuilder will favor a local driver if a driver service is set.

The current logic creates a RemoteWebDriver instance if a service is provided. `getLocalDriver()` is limited to only processing provided options.

Hmm, actually, I have an idea I want to try out based on this conversation... Let me throw something together and see if it works.

**Comment by titusfortner** [comment]
Ok, so I don't understand the purpose of [WebDriverInfo](https://github.com/SeleniumHQ/selenium/blob/selenium-4.7.0/java/src/org/openqa/selenium/WebDriverInfo.java#L27) abstraction for `getLocalDriver()` in [RemoteWebDriverBuilder](https://github.com/SeleniumHQ/selenium/blob/selenium-4.7.0/java/src/org/openqa/selenium/remote/RemoteWebDriverBuilder.java#L318).
 
Would it make sense to add a service constructor to it?

```
  Optional<WebDriver> createDriver(DriverService service, Capabilities capabilities) throws SessionNotCreatedException;
```

Or is there a better way for us to use the service class and return a local driver instance in that method instead of forcing it through `getRemoteDriver()`.

I'm thinking about a combination of:

1. Add default service instances to the subclassed `builder()` methods (e.g., Chrome - https://github.com/SeleniumHQ/selenium/blob/selenium-4.7.0/java/src/org/openqa/selenium/chrome/ChromeDriver.java#L87-L89)
```
public static RemoteWebDriverBuilder builder() {
    return RemoteWebDriver.builder().oneOf(new ChromeOptions())
        .withDriverService(ChromeDriverService.createDefaultService());
}
```

2. Change the [`getLocalDriver()` conditional](https://github.com/SeleniumHQ/selenium/blob/selenium-4.7.0/java/src/org/openqa/selenium/remote/RemoteWebDriverBuilder.java#L319) to allow service but not config of any kind

```
    if (remoteHost != null || clientConfig != null ) {
      return null;
    }
```

3. Allow local driver to be created with a provided service instance. (not sure how to do this part)

**Comment by code-with-abdullah** [comment]
Please look into this [Pull Request](https://github.com/SeleniumHQ/selenium/pull/11532). 
**I have proposed a very simple solution that would mitigate this issue.**

**Comment by diemol** [comment]
@titusfortner what I meant is that a local driver will be used if the service is set, else the code will continue to create a `RemoteWebDriver`. I guess I should put my thoughts in a PR, maybe it is easier to explain like that.

**Comment by krmahadevan** [comment]
@titusfortner @diemol - Can you please check if this is something that can be done for fixing this ?

https://github.com/SeleniumHQ/selenium/pull/11537

This PR is not ready to be merged because I have just disabled a test that checks if a client config was specified for a local driver and if yes, then throw an exception and bail out.

It still needs a test that checks if the client config when specified for a local driver is honoured.

The flip side of this PR is basically that, when a user needs to specify a client config to a local driver, they would need to work with `WebDriver` interface and not with an actual driver flavor because the underlying driver after creation is augmented post creation.

A code sample would look like below

```java
   @Test
  public void hello() {
    WebDriver driver = null;
    try {
      ClientConfig cfg = ClientConfig.defaultConfig().readTimeout(Duration.ofMinutes(1))
        .connectionTimeout(Duration.ofMinutes(1));
      driver = RemoteWebDriver.builder()
        .config(cfg)
        .oneOf(new ChromeOptions())
        .build();
    driver.get("https://www.google.com");
    } catch (Exception e) {
      if (driver != null) {
        driver.quit();
      }
    }
  }
```



@code-with-abdullah - FYI

**Comment by krmahadevan** [comment]
Ok, please ignore. That was a rush! That wont work. There's a lot to it than what meets the eye :)

**Comment by krmahadevan** [comment]
@diemol @titusfortner - I believe that there are two parts to this issue.

1. Not able to set read/connection timeouts when using local drivers (No Grid is involved). 
2. Not able to set read/connection timeouts when using the Grid via RemoteWebdriver.

I understand that we are still debating on how should the API look like for (1), but for (2) the only way to address it as far as I can think of is via JVM arguments. To address (2) I have raised a PR https://github.com/SeleniumHQ/selenium/pull/11538

This PR is intended for those users who are using the RemoteWebDriverBuilder in their tests and setting valid client read/connection timeouts at their end and are using the Selenium Grid. As far as I can tell, the RemoteWebDriver is NOT serialising and sending across the timeout values to the remote node, which would get parsed by one of the implementations of `WebDriverInfo` and then eventually gets translated into a local driver at the remote node end. The only way in that case is to be able to pass in the read and connection timeout values via JVM arguments.

Please help take a look at this PR.

If #11538 gets merged, then it would also address this defect partially as in, it provides a global way to alter these timeouts but a user will still NOT be able to alter these timeouts on a per WebDriver instantiation basis (Assuming that they have different webdriver instantiation mechanisms for different types of scenarios)

**Comment by titusfortner** [comment]
Number 2 is available using `RemoteWebDriverBuilder.withConfig()`

The problems include:
1. There is no constructor for local drivers accepting a config, which we could potentially do with #11532
2. While a driver can be started locally with the `RemoteWebDriverBuilder`, it is returned as an augmented `RemoteWebDriver` instance, which isn't ideal

I agree with Diego that we want to avoid System Property usage as much as possible unless absolutely necessary for the grid.

**Comment by krmahadevan** [comment]
@titusfortner - Here's a draft PR that aims at addressing (2) https://github.com/SeleniumHQ/selenium/pull/11545

**Callouts:**
* It does not use the RemoteWebDriverBuilder
* It introduces a new builder that is specific to local drivers

I chose not to use the existing RemoteWebDriverBuilder because 

* it seems that it was built to be able to primarily toggle between either this or that browser (for e.g., i have a scenario wherein I dont know if I am gonna get a IE or Chrome or Firefox node, and am really not bothered about the browser flavor i just need my test to run against any available browser sort of construct).

* A user can invoke one of the buillders from the driver classes, and still can try and point it to a remote node (Standalone or a grid) and execute it. For the remotewebdriver builder to work for local drivers, we would have to set the driverservice (in some manner) and this conflicts with the use case wherein I am building a ChromeDriver using the RemoteWebDriver and would like to run it against the grid.

* When working with local webdrivers, I don't really need it to be augmented and I need it to give me back a concrete webdriver implementation and NOT an interface which is what the RemoteWebDriver does.

Please help take a look at this PR and see if this aligns with what we are trying to do.

It primarily differs from the PR that @code-with-abdullah had raised via https://github.com/SeleniumHQ/selenium/pull/11532 in that, I have kept the changes confined ONLY to the CommandExecutor, because that's the only piece that is interested in a ClientConfig.

**Comment by krmahadevan** [comment]
@diemol - Soliciting your inputs as well here.

**Comment by diemol** [comment]
@krmahadevan I left comments in the PR. I think overall it looks good, however it'd be nice to help the person who submitted the initial PR with the idea of having an extra constructor.

**Comment by krmahadevan** [comment]
> @krmahadevan I left comments in the PR. I think overall it looks good, however it'd be nice to help the person who submitted the initial PR with the idea of having an extra constructor.

Absolutely @diemol . That's the way to encourage everyone to be part of Selenium. My PR will be in Draft only and I will get rid of it, once the original PR submitted by one of our users gets merged.

**Comment by TolstoyDotCom** [comment]
#11545 seems to work, at least for Chrome. I'd prefer to use FF, but it had an issue.

I opened the gitpod link on this project's homepage and ran this:
```
git fetch origin pull/11545/head:cktest
git checkout cktest
bazel build java/...
rake maven-install

```
I then downloaded ~/.m2 from gitpod and incorporated it into my local maven repo. I was then able to build my [censorship checker](https://github.com/TolstoyDotCom/more-speech) by setting the Selenium version to 4.7.2. However, now that I'm using Chrome I don't have the HAR extension so I need to find another way to get Twitter's AJAX. I'm looking into overriding XMLHttpRequest in Javascript to get that.

Also, if the issue with rolling this out is the constructors, why not hide those in a factory that would create all kinds of webdrivers? You'd consolidate all the possible settings into two types of classes: one that has generic settings (like the proxy to use, the grid URL, etc), and a series of classes that would have the platform-specific settings. It seems like, when using grid, you can request FF if it's available but use Chrome if it isn't, etc. That'd be something like this:

```
enum DriverBrowserType { CHROME, FIREFOX,...}

interface IDriverServiceFactory {
	IDriverConfig createConfig();
	IDriverPlatformConfig createPlatformConfig( DriverBrowserType driverBrowserType );
	IDriverService createDriverService( IDriverConfig driverConfig );
}

IDriverConfig driverConfig = factory.createConfig();
driverConfig.setProxy( javaProxy );
driverConfig.setSeleniumProxy( seleniumProxy );
driverConfig.setReadTimeout( duration );
driverConfig.setScriptTimeout( duration );
driverConfig.setGridInfo( gridInfo );		// if set, uses a grid browser. otherwise, uses a local browser.
// other settings

IDriverFirefoxConfig firefoxConfig = (IDriverFirefoxConfig) factory.createPlatformConfig( DriverBrowserType.FIREFOX );
firefoxConfig.setSomePlatformSpecificValue( 1 );
// other settings

driverConfig.addPlatformConfig( firefoxConfig );

IDriverChromeConfig chromeConfig = (IDriverChromeConfig) factory.createPlatformConfig( DriverBrowserType.CHROME );
chromeConfig.setSomePlatformSpecificValue( 2 );
// other settings

driverConfig.addPlatformConfig( chromeConfig );

IDriverService driverService = factory.createDriverService( driverConfig );

IDriver driver = driverService.createDriver();

driver.get( "http://localhost" );
...

```

**Comment by titusfortner** [comment]
@TolstoyDotCom it sounds like you want to use `RemoteWebDriverBuilder` which is intended for remote execution with a grid and supports things like passing multiple options instances with "first match" logic.

**Comment by TolstoyDotCom** [comment]
I'm the OP and the only thing I care about is opening the browser on my local machine. I'm the author of a desktop app that collects stats on how extensive Twitter censorship is. This is a very timely issue since censorship there has gotten even worse after Musk took over and I'd like to have data to back that up.

Using the steps I discussed above, I've verified that the PR resolves the issue of my script timing out, which was what I filed this issue over. However, I also need the AJAX messages that Twitter sends and so far I haven't been able get those using Chrome. I can get them using Firefox, but for some reason FF is now just hanging (for my timeout value). I was getting those using a proxy, but now that isn't working. Hopefully it won't take much more time before I can get one of those two working.

In any case, the code at the end of my last comment was a proposed changed to Selenium that I think would make it easier to deal with. The user wouldn't see RemoteWebDriver any more. In the general config the user would indicate whether they want a local or remote driver using either `setGridInfo` or `setUseRemote`.

What's in my last comment is what my proposed change would look like if the user wants a remote driver that's either Firefox or Chrome. The below is what my proposed change would look like if the user wants a local driver that's Firefox. If you don't like the idea of `createConfig` returning a mutable object, then that could be changed to `createConfigBuilder` and it would return an IDriverConfigBuilder object that would return a new, immutable IDriverConfig implementation for each setter.

```
// factory is an IDriverServiceFactory implementation
IDriverConfig driverConfig = factory.createConfig();
driverConfig.setProxy( javaProxy );
driverConfig.setSeleniumProxy( seleniumProxy );
driverConfig.setReadTimeout( duration );
driverConfig.setScriptTimeout( duration );
driverConfig.setGridInfo( null );	//	or driverConfig.setUseRemote( false );
// other settings

IDriverFirefoxConfig firefoxConfig = (IDriverFirefoxConfig) factory.createPlatformConfig( DriverBrowserType.FIREFOX );
firefoxConfig.setSomePlatformSpecificValue( 1 );
// other settings

driverConfig.addPlatformConfig( firefoxConfig );

IDriverService driverService = factory.createDriverService( driverConfig );

IDriver driver = driverService.createDriver();

driver.get( "http://localhost" );
...

```

**Comment by titusfortner** [comment]
We don't want to add more ways to do things that are already supported, and keeping Service concerns and Options concerns and Command Executor/HTTP Client concerns all separate makes more sense than implementing a new way to combine them all. It should be possible to implement this abstraction in your own code if you prefer that pattern.

It sounds like the read timeout wasn't actually the issue and that it is hanging for other reasons. If you are using a proxy with authentication, that isn't supported with geckodriver — https://github.com/mozilla/geckodriver/issues/1872
You might need to use a proxy-in-the-middle for working with credentials.

**Comment by diemol** [comment]
Closed via #11532
